### PR TITLE
Bevm calls byzcoin

### LINKED ISSUE #2201: Convert BEVM-Events to Instructions
In order to have a link from BEVM to ByzCoin, the following could be done:

- Add `executeInstruction` to `GlobalState`
- create new darc-identity `bevm_instr` with evm-contractID
- call it from BEVM upon receiving corresponding Event

1st use-case:
- update `ValueInstance` in ByzCoin with a BEVM contract

### PR DESCRIPTION
**Pull request by cgrigis**
**What this PR does**

This PR handles some events generated by the execution of an EVM contract by `bevm` and generates instructions allowing to spawn, invoke or delete a `byzcoin` contract. See #2187 and #2201 for more details regarding the motivation and the design.

The new generated instructions are signed by a new identity `evm_contract`, where the signature is just the address of the EVM contract. DARC rules must be defined using this identity to give authorization to the desired contracts.

Closes #2201

---

🙅‍ Friendly checklist:

- [X] 0. Code comments are added (or updated) when/where needed and explain the WHY of the code.
- [X] 1. Design choices, user documentation and any additional doc are added (or updated) in READMEs.
- [X] 2. Any new behaviour is tested and small units of code that can be are unit tested.
- [X] 3. Code comments are added on tests to explain what they do.
- [X] 4. Errors are systematically wrapped with a meaningful message using `xerrors.Errorf` and the `%v` verb.
- [X] 5. Hard limit of 80 chars is always respected.
- [X] 6. Changes are backward compatible.
- [X] 7. Indentation level does not exceed 5, although 4 is already suspicious.
- [X] 8. Functions, files, and packages are kept to a manageable size and decomposed into smaller units if needed.
- [X] 9. There are no magic values.

### CODE REVIEW

**Comment by cgrigis** [comment]
Documentation will follow in another PR (along with documentation for #2214).

File: bevm/bevm_call_byzcoin_test.go [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+	"go.dedis.ch/onet/v3"
+	"go.dedis.ch/onet/v3/log"
+	"golang.org/x/xerrors"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/stretchr/testify/require"
+)
+
+func init() {
+	err := byzcoin.RegisterGlobalContract(myValueContractID,
```
Comment (ineiti) [inline bevm/bevm_call_byzcoin_test.go]:
```suggestion
	log.ErrFatal(byzcoin.RegisterGlobalContract(myValueContractID,
```

Then you can remove the `if`-branch.

File: bevm/bevm_call_byzcoin_test.go [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+func (c myValueContract) Spawn(rst byzcoin.ReadOnlyStateTrie,
+	inst byzcoin.Instruction, cin []byzcoin.Coin) (sc []byzcoin.StateChange,
+	cout []byzcoin.Coin, err error) {
+	cout = cin
+
+	// Find the darcID for this instance.
+	var darcID darc.ID
+	_, _, _, darcID, err = rst.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
```
Comment (ineiti) [inline bevm/bevm_call_byzcoin_test.go]: add `fmt.Errorf("couldn't get darcID: %v", err)` here, and in other places.

File: bevm/bevm_call_byzcoin_test.go [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+			inst.InstanceID,
+			myValueContractID,
+			nil,
+			darcID),
+	}
+	return
+}
+
+// Inspired by byzcoin.Client.WaitProof()
+func waitProofGone(t *testing.T, cl *byzcoin.Client, interval time.Duration,
```
Comment (ineiti) [inline bevm/bevm_call_byzcoin_test.go]: Can you try to use `cl.UseNode(1)` in your tests and then remove the `waitProofGone` and other `WaitProof`s?

File: bevm/bevm_call_byzcoin_test.go [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+		}
+		time.Sleep(interval / 10)
+	}
+
+	return xerrors.Errorf("timeout reached and inclusion proof for %v "+
+		"still exists", id)
+}
+
+func Test_BEvmCallsByzcoin(t *testing.T) {
+	log.LLvl1("BEvmCallsByzcoin")
```
Comment (ineiti) [inline bevm/bevm_call_byzcoin_test.go]: remove debugging output - here and in other places.

File: byzcoin/transaction.go [hunk 5]
```diff
@@ -359,6 +359,26 @@ func (instr Instruction) Verify(st ReadOnlyStateTrie, msg []byte) error {
...
 }
 
+func (instr Instruction) usesForbiddenIdentities() bool {
+	// A synthetic instruction is currently not restricted
+	if instr.synthetic {
+		return false
+	}
+
+	// A genuine instruction cannot use an EVM contract identity
+	evmContractType := darc.Identity{
```
Comment (ineiti) [inline byzcoin/transaction.go]: For later: we should have constants for the types, not just `1`, `2`, ...

File: bevm/bevm_call_byzcoin_test.go (hunk 3 above)
Comment (cgrigis) [reply bevm/bevm_call_byzcoin_test.go]: > Can you try to use `cl.UseNode(1)` in your tests and then remove the `waitProofGone` and other `WaitProof`s?

How do I then check that an InstanceID exists (or no longer does)? Directly with `GetProof()`/`InclusionProof.Exists()`?

File: byzcoin/transaction.go (hunk 5 above)
Comment (cgrigis) [reply byzcoin/transaction.go]: Indeed. I resisted the urge to change that in this PR, trying to stay focused ;-) .

File: bevm/bevm_call_byzcoin_test.go (hunk 3 above)
Comment (ineiti) [reply bevm/bevm_call_byzcoin_test.go]: yes, that should work. The problem that `waitProofGone` and `WaitProof` solve is the communication with different nodes: if you send your transaction to node A, and get the proof from node B, you're not sure they are synchronized.

`cl.UseNode(1)` puts all communication through the same node. So if you send a transaction and it returns, you can be sure that the block has been applied in node 1.

File: bevm/bevm_call_byzcoin_test.go [hunk 6]
```diff
@@ -257,8 +258,7 @@ func Test_BEvmCallsByzcoin(t *testing.T) {
 	valID := byzcoin.ComputeNewInstanceID(myValueContractID, seed)
 
 	// Check that the new instance exists and holds the correct value
-	_, err = cl.WaitProof(valID, 10*time.Second, initValue)
-	require.NoError(t, err)
+	require.True(t, instanceIDExists(t, cl, valID, initValue))
```
Comment (ineiti) [inline bevm/bevm_call_byzcoin_test.go]: OK - I didn't think you want to check the value. So you could do this and wouldn't need an external method:

```
_, err = cl.WaitProof(valID, 0, initValue)
require.NoError(t, err)
```

**Pull request merged**